### ISSUE TITLE
Duplicate a Studentquiz activity also duplicates questions and this is not what teachers hope

### ISSUE THREAD

**Issue by zabellemotte** [frankkoch/moodle-mod_studentquiz#207]
Hello, 

We have been installing your very interesting Moodle activity plugin and some professors begin experimenting it with theirs students. We are amazed that duplicating a Studentquiz activity duplicates questions. It does not point toward the same questions like a Moodle test do but it actually create new questions in the databank that are exact copies of the existing questions.
We do not see any interest to such duplications. 
**Professors processing duplication of a Studentquiz activity are rather waiting for the creation of an empty Studentquiz activity**, with no question, but with the same configuration parameters.

Is it possible to implement it in a future version ?

Kisses

Zabelle

**Comment by frankkoch** [comment]
Hi Zabelle, as an easy workaround to get what you want:

Just create a new (of course empty) StudentQuiz activity with your prefered configuration and duplicate from there.

I leave this open for some time, maybe someone else wants to comment.

**Comment by dionysius** [comment]
Hi @zabellemotte 

TL;DR: It's highly possible that this can't be changed due to technical reasons.

TL;DR 2: I've changed my mind: I won't be able to fulfill the request as the hacks/workarounds have unforeseeable and unmaintainable implications in the future.

So if you want an empty StudentQuiz with the same settings, I have some things to note as well:
- After duplication, when in the question overview, you have a link/button to show all questions. Select them all and delete them. Now you have an empty StudentQuiz with exact the same settings with like 3 extra clicks.
- Or you might have not noticed, most of the global plugin settings of StudentQuiz act as default for new StudentQuizzes, thus are inherited from there. Talk to your moodle administrator to discuss these defaults maybe.

I was also investigating if your request can be fulfilled. *I wrote down a more extended explanation, since I wanted to link and try to discuss this issue with moodle developers, but I already found out, that there will be no satisfying solution:*

I've never used that feature yet so I was actually wondering what moodle actually does. According to https://docs.moodle.org/38/en/Capabilities/moodle/backup:backuptargetimport
> Together with the capability moodle/restore:restoretargetimport, this allows a user to duplicate an activity or resource [...]

This sounds like it does a "fake" backup and import and just gives the new activity another name. In the underlying code of the backup procedure, questions  **and assigned question categories** are part of that backup in StudentQuiz, but this is also the case for e.g. Quiz Module. The difference however is that the questions in StudentQuiz are in a question category of that StudentQuiz activity context, while the other questions - you are used to - are saved in categories of the course context. 

This context is required so we can give students permissions to create questions - we don't want students to be able to create/delete questions in the whole course, do we? More info about [context](https://docs.moodle.org/38/en/Context). Now technically, thinking about that, the duplication happens when Moodle tries to import: As I said earlier, StudentQuiz questions are in the question category of the context in this StudentQuiz activity, since the context is new, all these questions are considered new, since their original can't be found. In a e.g. Quiz import however, it can find their counterpart question category (the course question category) and the questions, since the context of the course can be found and these questions still exist in that context.

To add: Questions are added to the backup using the provided library features of Moodle, so we obey the coding standard of moodle. The only chance we have, is, if we can detect that this backup and/or restore code is called because of a duplication request and would somehow exclude/delete the questions in the backup/after the restore - but this sounds like a hack which can break any time in the future and no moodle developer expects this behaviour from a plugin.

Think of each StudentQuiz activity as a sandbox: Thanks to the context, we can give students permissions (without letting these permissions outbreak) they otherwise can't have, for the price, that everything inside this sandbox is part of it.

**Comment by zabellemotte** [comment]
Thanks for answers and feedback.
I know it is possible to suppress the questions in the duplicated activity. 
But this way of doing is not the used way Moodle works.
And professors are afraid because they do not know if questions will completely disappear from the course or not. 

I do not understand what is the pedagogical interest to have a duplicated activity with already posted questions in it. If someone duplicates an activity, it is to have an empty activity.
In Moodle, the default duplicate action makes a backup and restore without user data.
Please test an assign duplication with students submissions. You get a perfect copy of the assign without any student submission.
In the restore function, it is possible to define what data are user data and these data are not imported while duplication.

I can understand it is not easy to implement.
The duplication action is not well documented (I searched the user and dev documentation) and as far as you want to be able to export questions in the course question bank, I understand it is tricky.  
But please, understand the way the duplicate works in you plugin is unusual and confusing.

I had a look at Moodle permissions to see if it would be possible to disable the duplicate action on studentquiz activities and it is not possible. If I disable the restore action to some user roles, it will be for all Moodle activites. This is not wanted.

So, we will live with this difficulty for the moment, and I am sure this won't avoid us to have very interesting pedagogical returns about this plugin ...

Kisses

Zabelle

**Comment by dionysius** [comment]
I can't speak for the usage of Moodle. I only know (or could find out) how the internals of Moodle work.

> But this way of doing is not the used way Moodle works.
> [...]
> But please, understand the way the duplicate works in you plugin is unusual and confusing.

Unfortunately not true: this is exactly how Moodle works. If you put a question in a Quiz Activity in the category of said Quiz module (instead of course level), and duplicate the quiz, you'll end up with a copy for that question as well. (I just tested that in Moodle 3.8 to be sure I don't say something wrong)

The only difference is the (missing) documentation, that in StudentQuiz questions are tied to the it's category by default, where in Quiz they are in course level. In the question create/edit form you can select which category this question goes, only the defaults are different.
